### PR TITLE
581 forgot spec compliance

### LINKED ISSUE #581: Support properties and functionality for /forgot with JSON requests
per framework spec [here](https://github.com/stormpath/stormpath-framework-spec/blob/master/password-reset.md).

- Change `stormpath.web.forgot.*` to `stormpath.web.forgotPassword.*`
- Add support for `application/json` requests
- Ensure that the current HTML implementation conforms to the framework spec

### PR DESCRIPTION
**Pull request by mzumbadoate**
resolves #581

### CODE REVIEW

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java [hunk 1]
```diff
@@ -122,11 +156,22 @@ protected ViewModel onValidSubmit(HttpServletRequest request, HttpServletRespons
             //404 == resource does not exist.  Do not let the user know that the account does not
             //exist, otherwise we open up to phishing attacks
             if (e.getCode() != 404) {
+                if (UserAgents.get(request).isJsonPreferred()) {
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java]: This should be using `ContentNegotiationReslver`

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java [hunk 2]
```diff
@@ -122,11 +156,22 @@ protected ViewModel onValidSubmit(HttpServletRequest request, HttpServletRespons
...
+                    model.put("status", 200);
+                    model.put("message", "OK");
+                    return new DefaultViewModel("stormpathJsonView", model);
+                }
                 throw e;
             }
             //otherwise don't do anything
         }
 
+        if (UserAgents.get(request).isJsonPreferred()) {
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java]: This should be using `ContentNegotiationReslver`

File: extensions/servlet/src/main/resources/META-INF/resources/WEB-INF/jsp/stormpath/forgot-password.jsp [hunk 3]
```diff
@@ -39,6 +39,13 @@
                                 <p><sp:message key="stormpath.web.forgotPassword.form.instructions"/></p>
                             </div>
 
+                            <c:if test="${!empty status}">
```
Comment (mrioan) [inline extensions/servlet/src/main/resources/META-INF/resources/WEB-INF/jsp/stormpath/forgot-password.jsp]: I think we do not display error messages here. We do not validate the email address, we basically move to the `nextUri`. If I am not wrong you should remove this and related "error" logic/functionality

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java [hunk 4]
```diff
@@ -52,21 +55,51 @@ public ForgotPasswordController(Config config) {
...
+        return messageSource.getMessage(key, locale);
+    }
 
     @Override
     protected void appendModel(HttpServletRequest request, HttpServletResponse response, Form form, List<ErrorModel> errors,
                                Map<String, Object> model) {
         model.put("loginUri", loginUri);
+        String status = request.getParameter("status");
+        if ("invalid_sptoken".equals(status)) {
+            String key = "stormpath.web.forgotPassword.form.status";
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java]: If I am not wrong line 100 will end up being `stormpath.web.forgotPassword.form.status.status` (see that status is twice at the end). Does this really make sense?

File: extensions/servlet/src/main/resources/META-INF/resources/WEB-INF/jsp/stormpath/forgot-password.jsp [hunk 5]
```diff
@@ -39,6 +39,13 @@
...
+                                <div class="alert alert-dismissable alert-success">
+                                    <button type="button" class="close" data-dismiss="alert">&times;</button>
+                                    <p>${status}</p>
+                                </div>
+                            </c:if>
+
                             <c:if test="${!empty errors}">
```
Comment (mrioan) [inline extensions/servlet/src/main/resources/META-INF/resources/WEB-INF/jsp/stormpath/forgot-password.jsp]: I think we do not show errors any longer here as the email address is not validated. If i am right you can remove this "errors" block.

**Comment by mrioan** [comment]
Additionally:

## UAT

- *FAILED*: Spec [here](https://github.com/stormpath/stormpath-framework-spec/blob/master/password-reset.md) says: `There is no JSON response for the GET verb of this endpoint`. But if I do `curl -H "Accept: application/json" http://localhost:8080/forgot` I get a JSON Response.
- *FAILED*: It also states that `Regardless of whether or not the email address is associated with a user account, if the request prefers application/html, redirect the user to stormpath.web.forgotPassword.nextUri`. If I enter a wrong email I am kept in the same page and the text `Invalid email address.` is displayed


### Misc:

- The text "Email" and the field where the email address is typed should be horizontally aligned in the center. It seems they are aligned at the top now.

Back to you @mzumbadoate

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java (hunk 4 above)
Comment (mzumbadoate) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java]: the value os the variable status is "invalid_sptoken"
so actually the key will be in this case
stormpath.web.forgotPassword.form.status.invalid_sptoken

**Comment by mzumbadoate** [comment]
UAT's fixed

I see the "Email" label perfectly aligned to the input field

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java (hunk 4 above)
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ForgotPasswordController.java]: gotcha, thanks

**Comment by mrioan** [comment]
[This TCK test](https://github.com/stormpath/stormpath-framework-tck/blob/master/src/test/groovy/com/stormpath/tck/forgot/ForgotPasswordIT.groovy#L71) fails. When doing `curl -v -H "Accept: application/json" http://localhost:8080/forgot` you can see that the status of the response is `200`. TCK expect it to be different to `200` and `500`

File: examples/spring-security-spring-boot-webmvc/src/main/java/com/stormpath/spring/boot/examples/HelloController.java [hunk 6]
```diff
@@ -50,8 +50,12 @@ public String home(HttpServletRequest request, Model model) {
     }
 
     @RequestMapping("/restricted")
-    String restricted(HttpServletRequest request) {
-        return helloService.sayHello(AccountResolver.INSTANCE.getAccount(request));
+    String restricted(HttpServletRequest request, Model model) {
```
Comment (mrioan) [inline examples/spring-security-spring-boot-webmvc/src/main/java/com/stormpath/spring/boot/examples/HelloController.java]: What is this change for?

File: examples/spring-security-spring-boot-webmvc/src/main/java/com/stormpath/spring/boot/examples/RestrictedErrorController.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline examples/spring-security-spring-boot-webmvc/src/main/java/com/stormpath/spring/boot/examples/RestrictedErrorController.java]: 2016

File: examples/spring-security-spring-boot-webmvc/src/main/java/com/stormpath/spring/boot/examples/RestrictedErrorController.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.Map;
+
+/**
+ * @since 1.0.RC5
```
Comment (mrioan) [inline examples/spring-security-spring-boot-webmvc/src/main/java/com/stormpath/spring/boot/examples/RestrictedErrorController.java]: 1.0.0

File: examples/spring-security-spring-boot-webmvc/src/main/resources/templates/error.html [hunk 9]
```diff
@@ -0,0 +1,36 @@
+<!--/*
+~ Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline examples/spring-security-spring-boot-webmvc/src/main/resources/templates/error.html]: 2016

File: examples/spring-security-spring-boot-webmvc/src/main/resources/templates/fragments/head.html [hunk 10]
```diff
@@ -0,0 +1,50 @@
+<!--/*
+~ Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline examples/spring-security-spring-boot-webmvc/src/main/resources/templates/fragments/head.html]: 2016

File: examples/spring-security-webmvc/src/main/resources/application.properties [hunk 11]
```diff
@@ -17,4 +17,6 @@
 # If you have more than one application registered with Stormpath, you must specify which one
 # corresponds to this webapp:
 #
-#stormpath.application.href = https://api.stormpath.com/v1/applications/YOUR_APPLICATION_ID_HERE
\ No newline at end of file
+#stormpath.application.href = https://api.stormpath.com/v1/applications/YOUR_APPLICATION_ID_HERE
+
+#stormpath.web.csrf.token.enabled = false
```
Comment (mrioan) [inline examples/spring-security-webmvc/src/main/resources/application.properties]: please remove these lines

**Comment by mzumbadoate** [comment]
because the multiple commits actually duplicating functionality already in master I created a new branch and a new PR, #705

**Pull request closed without merging**